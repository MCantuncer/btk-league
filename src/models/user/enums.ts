import { registerEnumType } from 'type-graphql';

export enum UserTypeEnum {
  COACH = 1,
  PLAYER = 2,
  SPECTATOR = 3,
}

registerEnumType(UserTypeEnum, {
  name: 'UserTypeEnum',
  description: 'User Types',
});
