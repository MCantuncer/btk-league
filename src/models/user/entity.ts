import { Field, Int, ObjectType } from 'type-graphql';
import { getModelForClass, prop } from '@typegoose/typegoose';
import { UserTypeEnum } from './enums';
import { EntityWithIdAndDates } from '../base';

@ObjectType()
export class User extends EntityWithIdAndDates {
  @prop()
  @Field(() => UserTypeEnum)
  userType: UserTypeEnum;

  @prop()
  @Field(() => String)
  firstName: string;

  @prop()
  @Field(() => String)
  lastName: string;

  @prop()
  @Field(() => String)
  email: string;

  @prop()
  @Field(() => String)
  password: string;

  @prop()
  @Field(() => Int, { nullable: true })
  standing?: number | null;
}

export const UserModel = getModelForClass(User, {
  schemaOptions: { collection: 'users' },
});
