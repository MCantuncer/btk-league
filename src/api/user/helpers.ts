import { User } from '../../models/user/entity';
import * as jwt from 'jsonwebtoken';
import common from '../../../config/common';
import { ITokenData } from '../../utils/constants';

export class UserHelpers {
  static getToken(user: User, audience: string | undefined = undefined) {
    return jwt.sign(
      {
        id: user._id,
        email: user.email,
        firstName: user.firstName,
        lastName: user.lastName,
      } as ITokenData,
      common.jwt.secret,
      {
        algorithm: common.jwt.algorithm,
        subject: user.email,
        expiresIn: common.jwt.expiresIn,
        issuer: common.deploy_host,
        audience: audience,
      },
    );
  }

  static verifyToken(token: string): ITokenData | undefined {
    try {
      return jwt.verify(token, common.jwt.secret) as ITokenData;
    } catch (e) {
      return;
    }
  }
}
