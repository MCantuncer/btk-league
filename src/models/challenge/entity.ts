import { Field, ObjectType } from 'type-graphql';
import { EntityWithIdAndDates } from '../base';
import { getModelForClass, prop, Ref } from '@typegoose/typegoose';
import { User } from '../user/entity';

@ObjectType()
export class Challenge extends EntityWithIdAndDates {
  @prop({ ref: User, type: String })
  @Field(() => String)
  challenger: Ref<User>;

  @prop({ ref: User, type: String })
  @Field(() => String)
  challenged: Ref<User>;

  @prop()
  @Field(() => Boolean, { nullable: true })
  accepted?: boolean | null;
}

export const ChallengeModel = getModelForClass(Challenge, {
  schemaOptions: { collection: 'challenges' },
});
