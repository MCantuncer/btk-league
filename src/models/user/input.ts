import { Field, InputType } from 'type-graphql';
import { UserTypeEnum } from './enums';
import { IsEmail, MinLength } from 'class-validator';
import { BaseInputWithDate } from '../common';
import { ValidateEmailIsUnique } from '../../helpers/validators';

@InputType()
export class UserInput extends BaseInputWithDate {
  @Field(() => UserTypeEnum)
  userType: UserTypeEnum;

  @Field(() => String)
  firstName: string;

  @Field(() => String)
  lastName: string;

  @Field(() => String)
  @IsEmail()
  @ValidateEmailIsUnique()
  email: string;

  @Field(() => String)
  @MinLength(6)
  password: string;
}
