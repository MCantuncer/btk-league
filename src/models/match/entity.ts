import { Field, GraphQLTimestamp, Int, ObjectType } from 'type-graphql';
import { EntityWithIdAndDates } from '../base';
import { getModelForClass, modelOptions, prop, Ref } from '@typegoose/typegoose';
import { User } from '../user/entity';
import { WinnerEnum } from './enums';

@ObjectType()
@modelOptions({ schemaOptions: { _id: false } })
export class SetResult {
  @prop()
  @Field(() => Int)
  homeGame: number;

  @prop()
  @Field(() => Int)
  visitorGame: number;

  @prop()
  @Field(() => WinnerEnum, { nullable: true })
  winner?: WinnerEnum | null;
}

@ObjectType()
export class Match extends EntityWithIdAndDates {
  @prop({ ref: User, type: String })
  @Field(() => String)
  home: Ref<User>;

  @prop({ ref: User, type: String })
  @Field(() => String)
  visitor: Ref<User>;

  @prop()
  @Field(() => [SetResult], { nullable: true })
  result?: SetResult[] | null;

  @prop()
  @Field(() => WinnerEnum, { nullable: true })
  matchWinner?: WinnerEnum | null;

  @prop()
  @Field(() => GraphQLTimestamp, { nullable: true })
  matchDate?: Date | null;
}

export const MatchModel = getModelForClass(Match, {
  schemaOptions: { collection: 'matches' },
});
