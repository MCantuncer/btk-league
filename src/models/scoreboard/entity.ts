import { Field, Int, ObjectType } from 'type-graphql';
import { prop, Ref } from '@typegoose/typegoose';
import { User } from '../user/entity';

@ObjectType()
export class ScoreBoard {
  @prop({ ref: User, type: String })
  @Field(() => String)
  user: Ref<User>;

  @prop()
  @Field(() => Int)
  points: number;
}
