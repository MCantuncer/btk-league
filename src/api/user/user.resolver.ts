import { Arg, Ctx, Mutation, Query, Resolver } from 'type-graphql';
import { User, UserModel } from '../../models/user/entity';
import { UserInput } from '../../models/user/input';
import { copyFields } from '../../utils/dataUtils';
import { UserManager } from '../../models/user/manager';
import { verify } from 'argon2';
import { ApolloError } from 'apollo-server-express';
import { CheckResponse, LoginInput } from './models';

@Resolver(() => User)
export default class {
  @Query(() => User, { nullable: true })
  async me(@Ctx() { req }): Promise<User | null> {
    if (!req.session.userId) return null;

    return await UserModel.findOne({ _id: req.session.userId }).exec();
  }

  @Query(() => CheckResponse)
  async checkEmail(@Arg('email', () => String!) email: string): Promise<CheckResponse> {
    const user = await UserModel.findOne({ email }).lean();

    return {
      exists: !!user,
    };
  }

  @Mutation(() => User)
  async register(@Ctx() { req }, @Arg('input') input: UserInput): Promise<User | null> {
    copyFields(input, 'id', '_id');
    const user = await UserManager.registerUser(input);

    // const token = UserHelpers.getToken(user, common.deploy_host);
    // res.cookie(SESSION_COOKIE_NAME, token, { httpOnly: true, sameSite: true });
    req.session.userId = user.id;
    return user;
  }

  @Mutation(() => User)
  async login(@Ctx() { req, res }, @Arg('input') input: LoginInput): Promise<User | null> {
    const email = input.email;
    const user = await UserModel.findOne({ email: email }).exec();

    if (user && (await verify(user.password, input.password))) {
      // const token = UserHelpers.getToken(user, common.deploy_host);
      // res.cookie(SESSION_COOKIE_NAME, token, COOKIE_OPTIONS);
      req.session.userId = user.id;
      return user;
    } else {
      throw new ApolloError('Username or password is incorrect', 'USERNAME_OR_PASSWORD_IS_INCORRECT');
    }
  }
}
