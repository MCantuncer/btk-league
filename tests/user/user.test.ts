import { testMutate } from '../../jest.setup';
import { UserInputs, UserMutations } from './globals';
import { User } from '../../src/models/user/entity';

describe('User Tests', () => {
  it('Register a User', async () => {
    const graphqlResponse = await testMutate(UserMutations.Instance.SAVE, {
      input: { ...UserInputs.Instance().REGISTER_INPUT_1 },
    });

    const response = (graphqlResponse.data || {}) as { register: User };

    expect(response.register.email).toEqual(UserInputs.Instance().REGISTER_INPUT_1.email);
    expect(response).toMatchSnapshot();
  });
});
