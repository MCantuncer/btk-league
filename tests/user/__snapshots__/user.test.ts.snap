// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User Tests Check Email 1`] = `
Object {
  "checkEmail": Object {
    "exists": true,
  },
}
`;

exports[`User Tests Login User 1`] = `
Object {
  "login": Object {
    "email": "test@btk.com",
    "firstName": "Test",
    "id": "0bbdcc9f-a146-4d1b-a2bc-65257dadbd48",
    "lastName": "BTK",
    "userType": "PLAYER",
  },
}
`;

exports[`User Tests Register a User 1`] = `
Object {
  "register": Object {
    "email": "test@btk.com",
    "firstName": "Test",
    "id": "0bbdcc9f-a146-4d1b-a2bc-65257dadbd48",
    "lastName": "BTK",
    "userType": "PLAYER",
  },
}
`;
